src/core/arm7tdmi/alu.rs

```diff
@@ -123,14 +123,11 @@ impl Core {
     }
 
     pub fn lsr(&mut self, val: u32, amount: u32, _carry_in: bool, immediate: bool) -> u32 {
+        let amount = if immediate && amount == 0 { 32 } else { amount };
         match amount {
-            0 | 32 => {
-                if immediate {
-                    self.bs_carry_out = val.bit(31);
-                    0
-                } else {
-                    val
```

Comment:
So that was the bug? When `imm=0` and `amount=32` you'd return `val`, and now it's `0` instead?

Why is the behavior even different for immediates? (Non register?)